### ISSUE TITLE
@TypeName annotation scanner

### ISSUE THREAD

**Issue by bartoszwalacik** [javers/javers#263]
As a JaVers user I want to scan all my classes with @TypeName annotation (in given package)
with one call to `JaversBuilder` (preferably `scanTypeNames(String packageToScan)`) 
in order to use them in all kinds of JQL queries without getting TYPE_NAME_NOT_FOUND exception.

*@Since 1.4*

Vote here if you think that it could be useful.

**Comment by abibell** [comment]
**Symptom:** I still get the error.

I used [JaversSqlAutoConfiguration.java](https://github.com/javers/javers/blob/master/javers-spring-boot-starter-sql/src/main/java/org/javers/spring/boot/sql/JaversSqlAutoConfiguration.java) and added my `.scanTypeName(Employment.class).build();`. Upon debugging my `javers(ConnectionProvider connectionProvider)` doesn't get hit, however my other properties like `commitPropertiesProvider()` and `springSecurityAuthorProvider()` get hit.

Am I missing some configuration? I am using JaversSqlAutoConfiguration.java

```java
	
package com.test.config;
import java.util.HashMap;
import java.util.Map;

import javax.persistence.EntityManagerFactory;

import org.hibernate.SessionFactory;
import org.hibernate.dialect.Dialect;
import org.hibernate.engine.spi.SessionFactoryImplementor;
import org.javers.core.Javers;
import org.javers.core.MappingStyle;
import org.javers.core.diff.ListCompareAlgorithm;
import org.javers.hibernate.integration.HibernateUnproxyObjectAccessHook;
import org.javers.repository.sql.ConnectionProvider;
import org.javers.repository.sql.DialectName;
import org.javers.repository.sql.JaversSqlRepository;
import org.javers.repository.sql.SqlRepositoryBuilder;
import org.javers.spring.auditable.AuthorProvider;
import org.javers.spring.auditable.CommitPropertiesProvider;
import org.javers.spring.auditable.MockAuthorProvider;
import org.javers.spring.auditable.SpringSecurityAuthorProvider;
import org.javers.spring.auditable.aspect.JaversAuditableRepositoryAspect;
import org.javers.spring.boot.sql.DialectMapper;
import org.javers.spring.boot.sql.JaversProperties;
import org.javers.spring.jpa.JpaHibernateConnectionProvider;
import org.javers.spring.jpa.TransactionalJaversBuilder;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.autoconfigure.AutoConfigureAfter;
import org.springframework.boot.autoconfigure.condition.ConditionalOnClass;
import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingClass;
import org.springframework.boot.autoconfigure.orm.jpa.HibernateJpaAutoConfiguration;
import org.springframework.boot.autoconfigure.orm.jpa.JpaProperties;
import org.springframework.boot.context.properties.EnableConfigurationProperties;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.context.annotation.EnableAspectJAutoProxy;

import com.test.context.UserContext;
import com.test.context.UserContextUtil;
import com.test.model.Employment;

@Configuration
@EnableAspectJAutoProxy
@EnableConfigurationProperties(value = {JaversProperties.class,
		JpaProperties.class})
@AutoConfigureAfter(HibernateJpaAutoConfiguration.class)
public class JaversSqlAutoConfiguration {
	private static final Logger logger = LoggerFactory
			.getLogger(JaversSqlAutoConfiguration.class);

	private final DialectMapper dialectMapper = new DialectMapper();

	@Autowired
	private JaversProperties javersProperties;

	@Autowired
	private EntityManagerFactory entityManagerFactory;

	@Bean
	public DialectName javersSqlDialectName() {
		SessionFactoryImplementor sessionFactory = (SessionFactoryImplementor) entityManagerFactory
				.unwrap(SessionFactory.class);

		Dialect hibernateDialect = sessionFactory.getDialect();
		logger.info("detected Hibernate dialect: "
				+ hibernateDialect.getClass().getSimpleName());

		return dialectMapper.map(hibernateDialect);
	}

	@Bean
	public Javers javers(ConnectionProvider connectionProvider) {
		JaversSqlRepository sqlRepository = SqlRepositoryBuilder.sqlRepository()
				.withConnectionProvider(connectionProvider)
				.withDialect(javersSqlDialectName()).build();

		return TransactionalJaversBuilder.javers()
				.registerJaversRepository(sqlRepository)
				.withObjectAccessHook(new HibernateUnproxyObjectAccessHook())
				.withListCompareAlgorithm(ListCompareAlgorithm
						.valueOf(javersProperties.getAlgorithm().toUpperCase()))
				.withMappingStyle(MappingStyle.valueOf(
						javersProperties.getMappingStyle().toUpperCase()))
				.withNewObjectsSnapshot(javersProperties.isNewObjectSnapshot())
				.withPrettyPrint(javersProperties.isPrettyPrint())
				.withTypeSafeValues(javersProperties.isTypeSafeValues())
				.scanTypeName(Employment.class)
				.build();
	}

	@Bean(name = "authorProvider")
	@ConditionalOnMissingBean
	@ConditionalOnClass(name = {
			"org.springframework.security.core.context.SecurityContextHolder"})
	public AuthorProvider springSecurityAuthorProvider() {
		return new SpringSecurityAuthorProvider();
	}

	@Bean(name = "authorProvider")
	@ConditionalOnMissingBean
	@ConditionalOnMissingClass({
			"org.springframework.security.core.context.SecurityContextHolder"})
	public AuthorProvider unknownAuthorProvider() {
		return new MockAuthorProvider();
	}

	@Bean(name = "commitPropertiesProvider")
	@ConditionalOnMissingBean
	public CommitPropertiesProvider commitPropertiesProvider() {
		return new CommitPropertiesProvider() {
			@Override
			public Map<String, String> provide() {
				UserContext context = UserContextUtil.getContext();
				Map<String, String> map = new HashMap<String, String>();
				if (context.getEntityId() != null
						&& context.getGroupId() != null) {
					map.put("userId", context.getUserId().toString());
				}
				return map;
			}

		};
	}

	@Bean
	@ConditionalOnMissingBean
	public ConnectionProvider jpaConnectionProvider() {
		return new JpaHibernateConnectionProvider();
	}

	@Bean
	public JaversAuditableRepositoryAspect javersAuditableRepositoryAspect(
			Javers javers, AuthorProvider authorProvider,
			CommitPropertiesProvider commitPropertiesProvider) {
		return new JaversAuditableRepositoryAspect(javers, authorProvider,
				commitPropertiesProvider);
	}
}
```

**Comment by bartoszwalacik** [comment]
@abibell , this is not a good place for copy pasting long code listing, if you have questions about using `JaversBuilder.scanTypeName()` - ask on http://stackoverflow.com/questions/tagged/javers

**Comment by abibell** [comment]
1. I could bring in ann library and use it for `scanTypeNames()`. Are you ok with this?
2. What are your thoughts on using configuration property `packageToScan` which is baked in the base bean which will scan package if provided. It would not scan if there is no value provided (default). Package scanning maybe fairly common scenario like choice of algorithm and others to be considered for Configuration over customisation.

**Comment by bartoszwalacik** [comment]
1. some evaluation is required, maybe can you suggest some lib (there are many) and I will test it. It should be small and fast. We strive to keep javers-core dependencies list short. Maybe we will create a new module (named javers-core-ann-scanner) ?

2. yes, `packageToScan` property sounds like a good idea

**Comment by bartoszwalacik** [comment]
https://github.com/lukehutch/fast-classpath-scanner seems to have all buzzwords :)

**Comment by bartoszwalacik** [comment]
@abibell we could discuss details on javers chat https://gitter.im/javers/javers

**Comment by bartoszwalacik** [comment]
released in 2.3.0 see http://javers.org/documentation/features/#release-notes
